### PR TITLE
REF: standardize usage in DataFrame vs SparseDataFrame ops

### PR DESCRIPTION
**Pull request by jbrockmendel**
I _think_ that after this we're not far from being able to use the base class versions of _combine_frame, _combine_match_index, _combine_match_columns, and _combine_const.  That'll be a good day.

The only actual behavior changed is in SparseDataFrame._combine_match_columns, which is changed to match the other methods.  See #28025

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
> The only actual behavior changed is in SparseDataFrame._combine_match_columns, which is changed to match the other methods. See #28025

As mentioned in the issue: we could also leave the SparseDataFrame code alone as is (it's deprecated, so I don't think we should change behaviour, and I would personally also not put effort in cleaning it up), but still refactor the DataFrame ops (which would mean less sharing of code, for now, but I don't think that is a problem since the sparse code will be removed anyway, so then there will be no duplication anymore)

**Comment by jbrockmendel** [comment]
> (which would mean less sharing of code, for now, but I don't think that is a problem since the sparse code will be removed anyway, so then there will be no duplication anymore)

If I understand the suggestion correctly, this would involve copy/pasting all of the ops-based code that is currently shared.  Or do you have something else in mind?

**Comment by jorisvandenbossche** [comment]
> If I understand the suggestion correctly, this would involve copy/pasting all of the ops-based code that is currently shared. Or do you have something else in mind?

Yes, copy/paste all ops-based code that sparse needs and is currently shared with the other frame ops to the sparse module (and then we basically don't touch that code anymore until it is removed), and then you are free to refactor the ops code without needing to care about the deprecated sparse (and eg don't need to care about the small differences like https://github.com/pandas-dev/pandas/issues/28025)

(note this is just talk without looking at the actual code, I don't know if it is practically feasible. I just wouldn't care in this specific case about temporarily duplicating some code if that allows to refactor more easily the code that we are going to keep).

**Comment by jbrockmendel** [comment]
Thanks for clarifying.  Does the status quo for SparseDataFrame include bugfixes until it is removed?  If so, that would push towards continuing to share code.

**Comment by jorisvandenbossche** [comment]
>  Does the status quo for SparseDataFrame include bugfixes until it is removed?

Well, if somebody wants to do a PR for a bug fix, that is of course welcome (and non-shared code can still receive bugfixes). But if you ask *me* if I find it worth your time to do bug fixes for SparseDataFrame, then I would say no. Even if it would be benefiting from a generic bug fix that you do in the ops code, I don't think it is a problem that sparse wouldn't automatically get that bug fix anymore

**Comment by jbrockmendel** [comment]
I think I understand the suggestion, thanks.  Would you be OK with moving forward with the continue-sharing approach in this PR and in the next pass I'll try to use the approach you've described?

**Comment by jorisvandenbossche** [comment]
> Would you be OK with moving forward with the continue-sharing approach in this PR and in the next pass I'll try to use the approach you've described?

*If* we would go with a non-sharing appraoch, then I personally think we should do it already now and remove the sparse changes here. But that is still an "if" of course.
(I understood correctly that this PR was changing behaviour right?)

**Comment by jbrockmendel** [comment]
> (I understood correctly that this PR was changing behaviour right?)

This PR changes behavior, but we have not established that the behavior in question was intentional.

> then I personally think we should

I understand your preference.  My question is if the strength of that preference would make you -1 on moving forward with the less-preferred option (this) (conditional on resolving the behavior change)

**Comment by jorisvandenbossche** [comment]
>  My question is if the strength of that preference would make you -1 on moving forward with the less-preferred option (this) (conditional on resolving the behavior change)

Well, that depends on the potential behaviour change .. :)

> we have not established that the behavior in question was intentional.

In the issue, I asked about a code example, to be able to see what the change is in an example (and not in a theoretical explanation of the code that I can only understand by reading and understanding the code). But that is actually another reason to stop the sharing, then we wouldn't need to need to solve this question, we could just ignore the potential difference with SparseDataFrame, and use our mental energy for discussing more interesting things :-) 

To be clear, I don't really know what effort it would take to not share the code (I would think that, in this PR, it would mean reverting the changes in the sparse code, and copying (the original version of) some of the methods in that non-sparse code that you changed to the sparse code; but I didn't try it of course), so I do maybe a bit too lightly about it

**Comment by jbrockmendel** [comment]
@jreback if you're around, thoughts?  after more experimentation, im increasing convinced that a) the small behavior change is probably slightly more correct, b) it really doesn't matter, and c) it is possible to rip out a lot of the sparse ops code after implementing _construct_result

**Comment by jreback** [comment]
> @jreback if you're around, thoughts? after more experimentation, im increasing convinced that a) the small behavior change is probably slightly more correct, b) it really doesn't matter, and c) it is possible to rip out a lot of the sparse ops code after implementing _construct_result

agree on all points here; +1 on consolidating code and then simplifying ops. trying to isolate sparse via copy paste i think is just a mess and not worth any effort.

**Comment by jorisvandenbossche** [comment]
> im increasing convinced that a) the small behavior change is probably slightly more correct,

Then again, can you try to explain (show) the behaviour change?

**Comment by jbrockmendel** [comment]
> can you try to explain (show) the behaviour change?

```
df = pd.DataFrame([0])
sdf = df.to_sparse(fill_value=1)

# unchanged
>>> sdf.add(sdf[0], axis=0).default_fill_value
2.0

>>> (sdf + sdf[0]).default_fill_value
2.0  #  <-- PR
1    # <-- master
```

**Comment by TomAugspurger** [comment]
Essentially, binops propagate the default fill value.

Honestly, it

File: pandas/core/frame.py [hunk 1]
```diff
@@ -5314,38 +5313,50 @@ def _arith_op(left, right):
...
     def _combine_const(self, other, func):
         # scalar other or np.ndim(other) == 0
-        return ops.dispatch_to_series(self, other, func)
+        new_data = ops.dispatch_to_series(self, other, func)
+        return self._construct_result(other, new_data, func)
+
+    def _construct_result(self, other, result, func):
+        """
+        Compat for DataFrame/SparseDataFrame op result wrapping.
+
```
Comment (jreback) [inline pandas/core/frame.py]: can you add a full doc-string here (Paramaters / Returns)

File: pandas/core/frame.py [hunk 2]
```diff
@@ -5314,38 +5313,50 @@ def _arith_op(left, right):
...
+        `func` is included for compat with SparseDataFrame signature, is not
+        needed here.
+        """
+        out = self._constructor(result, index=self.index, copy=False)
+        # Pin columns instead of passing to constructor for compat with
+        #  non-unique columns case
+        out.columns = self.columns
+        return out
+        # TODO: finalize?  we do for SparseDataFrame
```
Comment (jreback) [inline pandas/core/frame.py]: finalize doesn't do really anything for frames, prob can't hurt though

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/frame.py]: especially in the frame+frame case it isn't obvious that is correct; I'd prefer to revisit

**Review by TomAugspurger — APPROVED** [review]
LGTM. Any other comments @jreback?

**Comment by jbrockmendel** [comment]
gentle ping.  This and #28268 are pseudo-blockers for implementing blockwise arithmetic.

(psuedo-blocker as in not actual blockers, but will make the implementation _much_ easier)

**Comment by jbrockmendel** [comment]
@jreback if we have to choose one more to get in more the weekend productivity streak ends, this should be it

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jbrockmendel 

@TomAugspurger you will likely need to rebase on the removals PR

**Pull request merged**